src/modules/auth/auth.service.ts

```diff
@@ -1,20 +1,118 @@
-import { Injectable, UnauthorizedException } from '@nestjs/common';
+import { BadRequestException, ConflictException, Injectable, UnauthorizedException } from '@nestjs/common';
 import { JwtService } from '@nestjs/jwt';
 
 import { Hash } from '@app/utils/hash.util';
 import { ConfigService } from '@nestjs/config';
 import { UsersService } from '@modules/user/user.service';
 import { User } from '@modules/user/user.entity';
 import { SigninDto } from '@modules/auth/dto/signin.dto';
+import cognito, {
+  CognitoIdentityProviderClient,
+  SignUpCommand,
+  ConfirmSignUpCommand,
+  InitiateAuthCommand,
+} from '@aws-sdk/client-cognito-identity-provider';
+import { RegisterRequestDto } from './dto/register.dto';
+import { checkUserExists } from '@app/utils/helper.util';
 
 @Injectable()
 export class AuthService {
+  private cognitoIdentity: CognitoIdentityProviderClient;
   constructor(
     private readonly jwtService: JwtService,
     private readonly configService: ConfigService,
     private readonly userService: UsersService,
   ) {}
 
+  async registerUser(registerDto: RegisterRequestDto) {
+    // check if user exists in cognito
+    const userExists = await checkUserExists(registerDto.email);
+    if (userExists.length > 0) {
+      throw new ConflictException('User already exists');
+    }
+    let attributes = [
+      { Name: 'name', Value: registerDto.name },
+      { Name: 'username', Value: registerDto.username },
+      { Name: 'email', Value: registerDto.email },
+    ];
+
+    try {
+      const input = {
+        ClientId: this.configService.get('USER_CLIENT_ID'),
+        Username: registerDto.username,
+        Password: registerDto.password,
+        UserAttributes: attributes,
+        ValidationData: attributes,
+      };
+      const signupCommand = new SignUpCommand(input);
+      const response = await this.cognitoIdentity.send(signupCommand);
+      const email = registerDto.email;
+      return {
+        statusCode: response.$metadata.httpStatusCode,
+        message: `User created successfully, check your email ${email} to confirm your account`,
+      };
+    } catch (error) {
+      const awsError = error as AWSError;
+      let message: string;
+      switch (awsError.name) {
+        case 'UsernameExistsException':
+          message = 'User already exists';
+          break;
+        case 'InvalidParameterException':
+          message = 'Invalid parameters provided';
+          break;
+        case 'UserNotFoundException':
+          break;
+        case 'TooManyRequestsException':
+          message = 'Too many requests, please try again later';
+          break;
+        default:
+          message = `An unexpected error occurred ${awsError.message}`;
+          break;
+      }
+      return { message: message, details: awsError };
```

Comment:
In case error happens, we should also send a message to the SNS incognito topic:
https://github.com/Carma-tech/Carma-tech-infra/blob/main/lib/lambda_functions/create_user.py#L45